File: src/main/java/shopping/member/infrastructure/MemberRepositoryImpl.java
package shopping.member.infrastructure;

import org.springframework.stereotype.Repository;
import shopping.member.domain.Member;
import shopping.member.domain.MemberRepository;

import java.util.Optional;

@Repository
public class MemberRepositoryImpl implements MemberRepository {
    private final MemberJpaRepository repository;

    public MemberRepositoryImpl(MemberJpaRepository repository) {
        this.repository = repository;
    }

    @Override
    public Member save(Member member) {
        return repository.save(MemberEntity.from(member)).toDomain();
    }

    @Override
    public Optional<Member> findByEmail(String email) {
        return repository.findByEmail(email)
                .map(MemberEntity::toDomain);
    }
}

File: src/main/kotlin/shopping/storage/db/MemberRepositoryImpl.kt
package shopping.storage.db

import org.springframework.stereotype.Repository
import shopping.domain.Member
import shopping.domain.MemberRepository

@Repository
class MemberRepositoryImpl(
    private val memberJpaRepository: MemberJpaRepository,
) : MemberRepository {
    override fun save(member: Member): Member {
        val entity = MemberEntity.from(member)
        val savedEntity = memberJpaRepository.save(entity)
        return savedEntity.toDomain()
    }

    override fun findByEmail(email: String): Member? {
        val entity =
            memberJpaRepository.findByEmail(email)
                ?: return null
        return entity.toDomain()
    }

    override fun existsByEmail(email: String): Boolean = memberJpaRepository.existsByEmail(email)
}
